keps/sig-network/3178-iptables-cleanup/README.md

```diff
@@ -0,0 +1,687 @@
+`OUTPUT` and `PREROUTING` chains of the `nat` table (because those are
+the only places you can call the `DNAT` target from, to redirect
+packets from a service IP to a pod IP), but neither masquerading nor
+dropping packets can occur at that point in IPTables processing:
+dropping packets can only occur from chains in the `filter` table,
+while masquerading can only occur from the `POSTROUTING` chain of the
+`nat` table (because the kernel can't know what IP to masquerade the
+packet to until after it has decided what interface it will send it
+out on, which necessarily can't happen until after it knows if it is
+going to DNAT it). To work around this, when Kubernetes wants to
+masquerade or drop a packet, it just marks it as needing to be
+masqueraded or dropped later, and then one of the later chains handles
+the actual masquerading/dropping.
+
+This approach was not _necessary_; kube-proxy could have just
+duplicated its matching logic between multiple IPTables chains, eg,
+having a rule saying "a packet sent to `192.168.0.11:80` should be
+redirected to `10.0.20.18:8080`" in one chain and a rule saying "a
+packet whose original pre-NAT destination was `192.168.0.11:80` should
+be masqueraded" in another.
+
+But using `KUBE-MARK-MASQ` allows kube-proxy to keep its logic more
+efficient, compact, and well-organized than it would be otherwise; it
+can just create a pair of adjacent rules saying "a packet sent to
+`192.168.0.11:80` should be redirected to `10.0.20.18:8080`, and
+should also be masqueraded".
+
+#### Eliminating `KUBE-MARK-DROP`
+
+In theory, `KUBE-MARK-DROP` is useful for the same reason as
+`KUBE-MARK-MASQ`, although in practice, it turns out to be much less
+necessary. Kube-proxy currently drops packets in two cases:
+
+  - When using the `LoadBalancerSourceRanges` feature on a service,
+    any packet to that service that comes from outside the valid
+    source IP ranges is dropped.
+
+  - When using `Local` external traffic policy, if a connection to a
+    service arrives on a node that has no local endpoints for that
+    service, it is dropped.
+
+In the `LoadBalancerSourceRanges` case, it would not be difficult to
+avoid using `KUBE-MARK-DROP`. The current rule logic creates a
+per-service `KUBE-FW-` chain that checks against each allowed source
+range and forwards traffic from those ranges to the `KUBE-SVC-` or
+`KUBE-XLB-` chain for the service. At the end of the chain, it calls
+`KUBE-MARK-DROP` on any unmatched packets.
+
+To do this without `KUBE-MARK-DROP`, we simply remove the
+`KUBE-MARK-DROP` call at the end of the `KUBE-FW-` chain, and add a
+new rule in the `filter` table, matching on the load balancer IP and
+port, and calling `-j DROP`. Any traffic that matched one of the
+allowed source ranges in the `KUBE-FW-` chain would have been DNAT'ed
+to a pod IP by this point, so any packet that still has the original
+load balancer IP as its destination when it reaches the `filter` table
```

Comment:
Caution - some folks access service LB's from within the cluster (and want us to stop internally redirecting!!).  If you blindly drop, this will not work, I think.

Reply:
OK, thinking about this some more, there really isn't much point in having the FW chain in `nat` anyway; we can just put all of the firewalling rules in the `filter` table and only the rewriting rules in the `nat` table.

(Note that firewalling currently happens before the ClusterIP redirect anyway, so it is already the case that if you want to connect to an LB IP from a pod, where the service uses LoadBalancerSourceRanges, then the source ranges need to include the pod IP. Our e2e tests require this behavior.)